sdk/tests/core/test_az_json.c

```
@@ -2978,6 +2981,264 @@ static void test_az_json_reader_chunked(void** state)
+    az_span destination_span = AZ_SPAN_FROM_BUFFER(destination);
+
+    int final_size;
+    az_result result = az_json_string_unescape(json, destination, 59, &final_size);
+
+    destination_span._internal.size = final_size;
+
+    assert_int_equal(AZ_OK, result);
+    assert_int_equal(expected._internal.size, final_size);
+    assert_string_equal(expected._internal.ptr, destination);
+    assert_true(az_span_is_content_equal(expected, destination_span));
+  }
+
+  // only escapes
+  {
+    az_span original = AZ_SPAN_FROM_STR("\\b\\f\\n\\r\\t\\\\");
+    az_span expected = AZ_SPAN_FROM_STR("\b\f\n\r\t\\");
+
+    char destination[59] = { 0 };
+    az_span destination_span = AZ_SPAN_FROM_BUFFER(destination);
+
+    int final_size;
+    az_result result = az_json_string_unescape(original, destination, 59, &final_size);
+
+    destination_span._internal.size = final_size;
+
+    assert_int_equal(AZ_OK, result);
+    assert_int_equal(expected._internal.size, final_size);
+    assert_string_equal(expected._internal.ptr, destination);
+    assert_true(az_span_is_content_equal(expected, destination_span));
+  }
+
+  // mix and match
+  {
+    az_span original
+        = AZ_SPAN_FROM_STR("Hello \\b My \\f Name \\n Is \\r Doctor \\t Green \\\\ Thumb");
+    az_span expected = AZ_SPAN_FROM_STR("Hello \b My \f Name \n Is \r Doctor \t Green \\ Thumb");
+
+    char destination[59] = { 0 };
+    az_span destination_span = AZ_SPAN_FROM_BUFFER(destination);
+
+    int final_size;
+    az_result result = az_json_string_unescape(original, destination, 59, &final_size);
+
+    destination_span._internal.size = final_size;
+
+    assert_int_equal(AZ_OK, result);
+    assert_int_equal(expected._internal.size, final_size);
+    assert_string_equal(expected._internal.ptr, destination);
+    assert_true(az_span_is_content_equal(expected, destination_span));
+  }
+
+  // fake escapes
+  {
+    az_span original = AZ_SPAN_FROM_STR("\\9");
+    az_span expected = AZ_SPAN_FROM_STR("\\9");
```

This is invalid JSON, and rather than copying the data verbatim, we should return an error `az_result`.

---

now i'm supposed to get into the json validation game   ? You yourself said that the customer can send whatever

---

> Also no other methos in this file validates that the input is a valid json

That is because all other APIs accept an `az_json_token`, which the user gets from a validating `az_json_reader`.

This new API is accepting an az_span, which can come from any arbitrary slice of input or byte array.

We can discuss how we want to think about validation of input, as a follow-up, separately.